cmd/head-main.go

```diff
@@ -0,0 +1,176 @@
+ */
+
+package cmd
+
+import (
+	"bufio"
+	"compress/bzip2"
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+	"syscall"
+
+	"golang.org/x/crypto/ssh/terminal"
+
+	"github.com/minio/cli"
+	"github.com/minio/mc/pkg/probe"
+)
+
+var (
+	headFlags = []cli.Flag{
+		cli.StringFlag{
+			Name:  "encrypt-key",
+			Usage: "decrypt object (using server-side encryption)",
+		},
+		cli.Int64Flag{
+			Name:  "n,lines",
+			Usage: "print the first 'n' lines",
+			Value: 10,
+		},
+	}
+)
+
+// Display contents of a file.
+var headCmd = cli.Command{
+	Name:   "head",
+	Usage:  "display first lines of an object",
+	Action: mainHead,
+	Before: setGlobalsFromContext,
+	Flags:  append(headFlags, globalFlags...),
+	CustomHelpTemplate: `NAME:
+  {{.HelpName}} - {{.Usage}}
+
+USAGE:
+  {{.HelpName}} [FLAGS] SOURCE [SOURCE...]
+
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}
+
+ENVIRONMENT VARIABLES:
+   MC_ENCRYPT_KEY:  List of comma delimited prefix=secret values
+EXAMPLES:
+   1. Display only first line from a compressed object on Amazon S3.
+      $ {{.HelpName}} -n 1 s3/csv-data/population.csv.gz
+
+   2. Display only first line from server encrypted object on Amazon S3.
+      $ {{.HelpName}} -n 1 --encrypt-key 's3/csv-data=32byteslongsecretkeymustbegiven1' s3/csv-data/population.csv
+`,
+}
+
+// headURL displays contents of a URL to stdout.
+func headURL(sourceURL string, encKeyDB map[string][]prefixSSEPair, nlines int64) *probe.Error {
+	var reader io.ReadCloser
+	switch sourceURL {
+	case "-":
+		reader = os.Stdin
+	default:
+		var err *probe.Error
+		var metadata map[string]string
+		if reader, metadata, err = getSourceStreamMetadataFromURL(sourceURL, encKeyDB); err != nil {
+			return err.Trace(sourceURL)
+		}
+		ctype := metadata["Content-Type"]
+		if strings.Contains(ctype, "gzip") {
+			var e error
+			reader, e = gzip.NewReader(reader)
+			if e != nil {
+				return probe.NewError(e)
+			}
+			defer reader.Close()
+		} else if strings.Contains(ctype, "bzip") {
+			defer reader.Close()
+			reader = ioutil.NopCloser(bzip2.NewReader(reader))
+		} else {
+			defer reader.Close()
+		}
+	}
+	return headOut(reader, nlines).Trace(sourceURL)
+}
+
+// headOut reads from reader stream and writes to stdout. Also check the length of the
+// read bytes against size parameter (if not -1) and return the appropriate error
+func headOut(r io.Reader, nlines int64) *probe.Error {
+	var e error
+	var stdout io.Writer
+
+	// In case of a user showing the object content in a terminal,
+	// avoid printing control and other bad characters to avoid
+	// terminal session corruption
+	if terminal.IsTerminal(int(os.Stdout.Fd())) {
+		stdout = newPrettyStdout(os.Stdout)
+	} else {
+		stdout = os.Stdout
+	}
+
+	// Initialize a new scanner.
+	scn := bufio.NewScanner(r)
+
+	// Negative number of lines means default number of lines.
+	if nlines < 0 {
+		nlines = 10
+	}
+
+	for scn.Scan() && nlines > 0 {
+		if _, e = stdout.Write(scn.Bytes()); e != nil {
```

Comment:
Suggestion: `var e error` need not be declared. Instead we can have error variable locally scoped inside the if block checking `stdout.Write`'s error value, similar to how we are checking `scn.Err` below.